---
title: Genetic Structure as Generative Poetic Form
author: J. Hagedorn
date: '2018-05-20'
tags:
  - genomics
  - prosody
slug: genetic-structure-as-generative-poetic-form
---

<script src="/rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<link href="/rmarkdown-libs/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="/rmarkdown-libs/datatables-binding/datatables.js"></script>
<script src="/rmarkdown-libs/jquery/jquery-3.6.0.min.js"></script>
<link href="/rmarkdown-libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="/rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="/rmarkdown-libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="/rmarkdown-libs/dt-ext-responsive/css/responsive.dataTables.min.css" rel="stylesheet" />
<script src="/rmarkdown-libs/dt-ext-responsive/js/dataTables.responsive.min.js"></script>
<link href="/rmarkdown-libs/dt-ext-buttons/css/buttons.dataTables.min.css" rel="stylesheet" />
<script src="/rmarkdown-libs/dt-ext-buttons/js/dataTables.buttons.min.js"></script>
<script src="/rmarkdown-libs/dt-ext-buttons/js/buttons.html5.min.js"></script>
<script src="/rmarkdown-libs/dt-ext-buttons/js/buttons.colVis.min.js"></script>
<script src="/rmarkdown-libs/dt-ext-buttons/js/buttons.print.min.js"></script>
<link href="/rmarkdown-libs/nouislider/jquery.nouislider.min.css" rel="stylesheet" />
<script src="/rmarkdown-libs/nouislider/jquery.nouislider.min.js"></script>
<link href="/rmarkdown-libs/selectize/selectize.bootstrap3.css" rel="stylesheet" />
<script src="/rmarkdown-libs/selectize/selectize.min.js"></script>
<link href="/rmarkdown-libs/crosstalk/css/crosstalk.min.css" rel="stylesheet" />
<script src="/rmarkdown-libs/crosstalk/js/crosstalk.min.js"></script>


<p>Part of the long poetic sequence I’m working on is a poem called ‘I’ (or maybe ‘I-I’), whose sections are each devoted to a theme related to a given gene, and structured based on the nucleotide sequence from that gene. In order to create these poetic patterns, we need actual genetic sequence data to which patterning algorithms can be applied.</p>
<p>Here, I’ll lay out the steps used to gather and create these patterns:</p>
<pre class="r"><code>library(tidyverse);library(DT); library(tidytext)
library(GO.db);library(hgu133plus2probe);
library(hgu133plus2.db)
genome &lt;- hgu133plus2probe </code></pre>
<div id="step-1-find-genes-that-perform-functions" class="section level2">
<h2>Step 1: Find Genes that Perform Functions</h2>
<p>The genome is vast, and starting at random at a given nucleotide seems unnecessarily naive, especially with the incredible amount of work that has been done in annotating genes based on their functions. This allows one to find specific sequences of nulceotides that are associated with a given biological function (<em>visual perception</em>, for instance) and to use those for relevant portions of the poem.</p>
<p>To query to Gene Ontology (GO) database, I use the <code>GO.db</code> package<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, which queries the <a href="http://amigo.geneontology.org/grebe"><em>AmiGo 2</em> database</a>. The code below defines a convenience function which takes a given search term string and returns a <code>tibble</code> with the related <code>GOID</code> and <code>GOTERM</code> fields, which can then be manually reviewed for interest. Once one or more terms are identified, these can be copied and manually added to the list in the next step which queries.</p>
<pre class="r"><code>go_tbl &lt;- toTable(GOTERM)[-c(1)]

search_go &lt;- function(search_string){
  output &lt;-
    go_tbl %&gt;%
    dplyr::filter(str_detect(Term,search_string)) %&gt;%
    dplyr::select(go_id,Term,Definition,Ontology) %&gt;%
    dplyr::distinct()
    
  return(output)
}</code></pre>
<p>Let’s define search terms for various topics of interest as regular expressions, which we can then pass to the database, and save these in a list called <code>searches</code>. Below are a handful of examples:</p>
<pre class="r"><code>searches &lt;- list()
searches$vision_search &lt;- search_go(&quot; vision|^vision|sight| visual|^visual|ocular&quot;)
searches$hearing_search &lt;- search_go(&quot;hearing| ear |^ear | auditory|^auditory&quot;)
searches$skin_search &lt;- search_go(&quot; skin |^skin| melanoma |^touch | touch &quot;)
searches$language_search &lt;- search_go(&quot; language |^langauge| speech |^speech&quot;)
searches$spine_search &lt;- search_go(&quot; spine |^spine&quot;)</code></pre>
<p>We can then review the resulting tables to find the genes which are of greatest interest for a given section of the poem, like so:</p>
<pre class="r"><code>searches$vision_search %&gt;%
  datatable(
    rownames = F, filter = &quot;top&quot;,
    caption = &#39;Summary of Gene Ontology Terms for Search&#39;,
    colnames = c(&#39;GO ID&#39;,&#39;GO Term&#39;,&#39;Definition&#39;,&#39;BP&#39;),
    extensions = c(&#39;Responsive&#39;,&#39;Buttons&#39;),
    options = list(
      pageLength = 5, lengthMenu = c(5, 15),
      dom = &#39;C&lt;&quot;clear&quot;&gt;lfrtip&#39;, buttons = c(&#39;colvis&#39;)
    )
  )</code></pre>
<div id="htmlwidget-1" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"filter":"top","vertical":false,"filterHTML":"<tr>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\" disabled=\"\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n<\/tr>","extensions":["Responsive","Buttons"],"caption":"<caption>Summary of Gene Ontology Terms for Search<\/caption>","data":[["GO:0001744","GO:0001748","GO:0002074","GO:0007601","GO:0007632","GO:0008542","GO:0014725","GO:0014726","GO:0014727","GO:0050908","GO:0051356","GO:0060006","GO:0060007","GO:0120303","GO:0150063"],["insect visual primordium formation","insect visual primordium development","extraocular skeletal muscle development","visual perception","visual behavior","visual learning","regulation of extraocular skeletal muscle development","negative regulation of extraocular skeletal muscle development","positive regulation of extraocular skeletal muscle development","detection of light stimulus involved in visual perception","visual perception involved in equilibrioception","angular vestibuloocular reflex","linear vestibuloocular reflex","visually-mediated background adaptation","visual system development"],["Establishment of the optic lobe placode. In Drosophila, for example, the placode appears in the dorsolateral region of the head in late stage 11 embryos and is the precursor to the larval visual system.","The process whose specific outcome is the progression of the optic placode over time, from its formation to the mature structure. During embryonic stage 12 the placode starts to invaginate, forming a pouch. Cells that will form Bolwig's organ segregate from the ventral lip of this pouch, remaining in the head epidermis. The remainder of the invagination loses contact with the outer surface and becomes the optic lobe. An example of this process is found in Drosophila melanogaster.","The process whose specific outcome is the progression of the extraocular skeletal muscle over time, from its formation to the mature structure. The extraocular muscle is derived from cranial mesoderm and controls eye movements. The muscle begins its development with the differentiation of the muscle cells and ends with the mature muscle. An example of this process is found in Mus musculus.","The series of events required for an organism to receive a visual stimulus, convert it to a molecular signal, and recognize and characterize the signal. Visual stimuli are detected in the form of photons and are processed to form an image.","The behavior of an organism in response to a visual stimulus.","Any process in an organism in which a change in behavior of an individual occurs in response to repeated exposure to a visual cue.","Any process that modulates the frequency, rate or extent of extraocular skeletal muscle development. Extraocular skeletal muscle development is the process whose specific outcome is the progression of the extraocular skeletal muscle over time, from its formation to the mature structure. The extraocular muscle is derived from cranial mesoderm and controls eye movements. The muscle begins its development with the differentiation of the muscle cells and ends with the mature muscle.","Any process that stops, prevents, or reduces the frequency, rate or extent of extraocular skeletal muscle development. Extraocular skeletal muscle development is the process whose specific outcome is the progression of the extraocular skeletal muscle over time, from its formation to the mature structure. The extraocular muscle is derived from cranial mesoderm and controls eye movements. The muscle begins its development with the differentiation of the muscle cells and ends with the mature muscle.","Any process that activates, maintains or increases the frequency, rate or extent of extraocular skeletal muscle development. Extraocular skeletal muscle development is the process whose specific outcome is the progression of the extraocular skeletal muscle over time, from its formation to the mature structure. The extraocular muscle is derived from cranial mesoderm and controls eye movements. The muscle begins its development with the differentiation of the muscle cells and ends with the mature muscle.","The series of events involved in visual perception in which a light stimulus is received and converted into a molecular signal.","The series of events during equilibrioception required for an organism to receive a visual stimulus, convert it to a molecular signal, and recognize and characterize the signal. Visual input plays an important role in the ability of an organism to perceive its orientation with respect to gravity.","A vestibular reflex by which a response to an angular acceleration stimulus begins with an afferent nerve impulse from a receptor in the semi-circular canal and ends with the compensatory action of eye muscles. Signaling never reaches a level of consciousness.","A vestibular reflex by which a response to a linear acceleration stimulus begins with an afferent nerve impulse from a receptor in the otolith and ends with the compensatory action of eye muscles. Signaling never reaches a level of consciousness.",null,null],["BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP","BP"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>GO ID<\/th>\n      <th>GO Term<\/th>\n      <th>Definition<\/th>\n      <th>BP<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"lengthMenu":[5,15],"dom":"C<\"clear\">lfrtip","buttons":["colvis"],"columnDefs":[],"order":[],"autoWidth":false,"orderClasses":false,"orderCellsTop":true,"responsive":true}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="step-2-subset-genome-based-on-functions-of-interest" class="section level2">
<h2>Step 2: Subset Genome based on Functions of Interest</h2>
<p>After reading through and doing some research on the gene ontology terms above, I select a specific term/function and then proceed to pull:</p>
<ol style="list-style-type: decimal">
<li><p>Relevant genes and annotation data from the human genome.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a></p></li>
<li><p>Probe data which includes the actual nucleotide sequences for each gene.<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a></p></li>
</ol>
<p>Below is a convenience function which takes a list of <code>GOID</code> strings as an argument and retrieves the related genes and functional annotation:</p>
<pre class="r"><code>get_genes &lt;- function(go_ids){
  
  output &lt;-
    hgu133plus2.db %&gt;%
    select(
      keytype = &quot;GO&quot;, 
      columns = c(
        &quot;GO&quot;,&quot;ONTOLOGY&quot;,&quot;EVIDENCE&quot;,&quot;SYMBOL&quot;,&quot;GENENAME&quot;,&quot;MAP&quot;,&quot;PATH&quot;,&quot;PROBEID&quot;
      ), 
      keys = go_ids
    ) %&gt;%
    mutate(
      EVIDENCE_TYPE = recode(
        EVIDENCE,
        `IMP` = &quot;inferred from mutant phenotype&quot;,
        `IGI` = &quot;inferred from genetic interaction&quot;,
        `IPI` = &quot;inferred from physical interaction&quot;,
        `ISS` = &quot;inferred from sequence similarity&quot;,
        `IDA` = &quot;inferred from direct assay&quot;,
        `IEP` = &quot;inferred from expression pattern&quot;,
        `IEA` = &quot;inferred from electronic annotation&quot;,
        `TAS` = &quot;traceable author statement&quot;,
        `NAS` = &quot;non-traceable author statement&quot;,
        `ND` = &quot;no biological data available&quot;,
        `IC` = &quot;inferred by curator&quot;,
        `IBA` = &quot;inferred from biological aspect of ancestor&quot;
      )
    ) %&gt;%
    mutate_if(
      .predicate = is.character,
      .funs = funs(as_factor)
    ) %&gt;%
    inner_join(genome, by = c(&quot;PROBEID&quot; = &quot;Probe.Set.Name&quot;)) %&gt;%
    inner_join(go_tbl[c(1:2)], by = c(&quot;GO&quot; = &quot;go_id&quot;)) %&gt;%
    dplyr::select(GENENAME,SYMBOL,sequence,MAP,PATH,PROBEID,x,y,GO,GOTERM = Term,ONTOLOGY,EVIDENCE,EVIDENCE_TYPE) %&gt;%
    distinct() 
  
  return(output)
  
}</code></pre>
<p>We can then filter the output by the strength of evidence or any other variable to select the specific gene and position that we want to use.<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a></p>
<pre class="r"><code># Use named lists for the purpose of documentation
vision_genes &lt;- 
  get_genes(
    go_ids = c(
      `visual perception` = &quot;GO:0007601&quot;, 
      `detection of light` = &quot;GO:0050908&quot;,
      `visual equilibrioception` = &quot;GO:0051356&quot;
    )
  )

hearing_genes &lt;-
  get_genes(
    go_ids = c(
      `response to auditory stimulus` = &quot;GO:0010996&quot;,
      `auditory behavior` = &quot;GO:0031223&quot;,
      `ear morphogenesis` = &quot;GO:0042471&quot;,
      `ear development` = &quot;GO:0043583&quot;,
      `inner ear cell fate commitment` = &quot;GO:0060120&quot;
    )
  )

skin_genes &lt;- 
  get_genes(
    go_ids = c(
      `establishment of skin barrier` = &quot;GO:0061436&quot;, 
      `skin development` = &quot;GO:0043588&quot;,
      `skin morphogenesis` = &quot;GO:0043589&quot;
    )
  )</code></pre>
</div>
<div id="step-3-select-a-specific-gene" class="section level2">
<h2>Step 3: Select a Specific Gene</h2>
<p>Out of all of the genes identified, we need to continue inward toward the needle in the haystack. This may include:</p>
<ul>
<li>Filtering by <a href="http://geneontology.org/page/guide-go-evidence-codes">evidence type</a></li>
<li>Determining which precise gene function is the best fit for the poem (e.g. for the <code>vision_genes</code>, these include visual perception, detection of light stimulus involved in visual perception…)</li>
</ul>
<p>Based on the final selection of gene, we pass the name of the gene from the <code>SYMBOL</code> field (e.g. <code>"RAX"</code>) to a convenience function which will return the poetic patterns available based on the gene’s nucleotide sequence and accompanying base pairs.</p>
<pre class="r"><code>get_patterns &lt;- function(gene_id){
  
  # Filter gene data based on selection
  gene_df &lt;-
    hgu133plus2.db %&gt;%
    select(
      keytype = &quot;SYMBOL&quot;, 
      columns = c(
        &quot;GO&quot;,&quot;ONTOLOGY&quot;,&quot;EVIDENCE&quot;,&quot;SYMBOL&quot;,&quot;GENENAME&quot;,&quot;MAP&quot;,&quot;PATH&quot;,&quot;PROBEID&quot;
      ), 
      keys = gene_id
    ) %&gt;%
    mutate(
      EVIDENCE_TYPE = recode(
        EVIDENCE,
        `IMP` = &quot;inferred from mutant phenotype&quot;,
        `IGI` = &quot;inferred from genetic interaction&quot;,
        `IPI` = &quot;inferred from physical interaction&quot;,
        `ISS` = &quot;inferred from sequence similarity&quot;,
        `IDA` = &quot;inferred from direct assay&quot;,
        `IEP` = &quot;inferred from expression pattern&quot;,
        `IEA` = &quot;inferred from electronic annotation&quot;,
        `TAS` = &quot;traceable author statement&quot;,
        `NAS` = &quot;non-traceable author statement&quot;,
        `ND` = &quot;no biological data available&quot;,
        `IC` = &quot;inferred by curator&quot;,
        `IBA` = &quot;inferred from biological aspect of ancestor&quot;
      )
    ) %&gt;%
    mutate_if(
      .predicate = is.character,
      .funs = list(~as_factor(.))
    ) %&gt;%
    inner_join(genome, by = c(&quot;PROBEID&quot; = &quot;Probe.Set.Name&quot;)) %&gt;%
    inner_join(go_tbl[c(1:2)], by = c(&quot;GO&quot; = &quot;go_id&quot;)) %&gt;%
    dplyr::select(GENENAME,SYMBOL,sequence,MAP,PATH,PROBEID,x,y,GO,GOTERM = Term,ONTOLOGY,EVIDENCE,EVIDENCE_TYPE) %&gt;%
    distinct() %&gt;%
    # Arrange by position
    arrange(x) %&gt;%
    dplyr::select(sequence) %&gt;%
    unnest_tokens(
      nucleotide,
      sequence,
      token =  &quot;character_shingles&quot;,
      n = 1
    ) %&gt;%
    # Define the other half of the base pair
    mutate(
      pair = recode(
        nucleotide,
        `a` = &quot;t&quot;,
        `t` = &quot;a&quot;,
        `c` = &quot;g&quot;,
        `g` = &quot;c&quot;
      )
    )
  
  # Define patterns
  tercet_purines &lt;-
    gene_df %&gt;%
    mutate(
      # Duplicate if purines
      nucleotide = recode(nucleotide, `a` = &quot;a|a&quot;,`g` = &quot;g|g&quot;),
      pair = recode(pair, `a` = &quot;a|a&quot;,`g` = &quot;g|g&quot;),
      base_pair = paste0(nucleotide,&quot;|&quot;,pair)
    ) %&gt;%
    summarize(pattern = base::paste(base_pair,collapse = &quot;||&quot;)) %&gt;%
    c()

  tercet_amino &lt;- 
    gene_df %&gt;%
    mutate(
      base_pair = paste0(nucleotide,pair)
    ) %&gt;%
    summarize(pattern = base::paste(base_pair,collapse = &quot;|&quot;)) %&gt;%
    # Add new | for every tercet break
    mutate(pattern = gsub(&#39;(.{9})&#39;, &#39;\\1|&#39;, pattern)) %&gt;%
    c()
  
  couplet_amino &lt;-
    gene_df %&gt;%
    mutate(
      row = row_number(),
      codon = cut(row, nrow(.) %/% 3, labels=FALSE)
    ) %&gt;%
    group_by(codon) %&gt;%
    summarize(
      nucleotide = base::paste(nucleotide,collapse = &quot;&quot;),
      pair = base::paste(pair,collapse = &quot;&quot;)
    ) %&gt;%
    mutate(
      pattern = base::paste0(nucleotide,&quot;|&quot;,pair)
    ) %&gt;%
    summarize(pattern = base::paste(pattern,collapse = &quot;||&quot;)) %&gt;%
    c()
  
  patterns &lt;- 
    list(
      `gene_df` = gene_df,
      `tercet_purines` = tercet_purines,
      `tercet_amino` = tercet_amino,
      `couplet_amino` = couplet_amino
    )
  
  return(patterns)
  
}</code></pre>
</div>
<div id="step-4-convert-gene-sequence-to-poetic-pattern" class="section level2">
<h2>Step 4: Convert Gene Sequence to Poetic Pattern</h2>
<p>Once we figure out which gene we want to use, we can use its <code>gene_id</code> to convert it into patterns using the <code>get_patterns()</code> function defined above:</p>
<pre class="r"><code>vision_patterns &lt;- get_patterns(gene_id = &quot;RAX&quot;)</code></pre>
<p>There are a few options to start with (and these will likely grow) for defining patterns for prosody. In each of these, a <code>|</code> denotes a line end and a <code>||</code> denotes a stanza end:</p>
<ul>
<li><code>tercet_purines</code>: a pattern in which each base pair is made into a tercet, with purines (adenine <code>a</code> and guanine <code>g</code>) getting 2 lines because they have 2 rings, and pyramidines (thymine <code>t</code> and cytosine <code>c</code>) getting one line because they have 1 ring. <em>Example: The base pair <code>at</code> would correspond to the tercet pattern <code>aat</code></em></li>
</ul>
<pre class="r"><code>str(vision_patterns$tercet_purines$pattern)</code></pre>
<pre><code>##  chr &quot;c|g|g||a|a|t||c|g|g||t|a|a||g|g|c||g|g|c||g|g|c||g|g|c||a|a|t||a|a|t||c|g|g||g|g|c||t|a|a||c|g|g||t|a|a||t|a|a|&quot;| __truncated__</code></pre>
<ul>
<li><code>tercet_amino</code>: a pattern in which every codon (i.e. 3 base pairs) corresponding to an amino acid gets its own tercet, where the initial nucleotide from the base pair corresponds to the start word of a line, and the other nucleotide to the end word of the line.</li>
</ul>
<pre class="r"><code>str(vision_patterns$tercet_amino$pattern)</code></pre>
<pre><code>##  chr &quot;cg|at|cg||ta|gc|gc||gc|gc|at||at|cg|gc||ta|cg|ta||ta|gc|gc||gc|at|ta||cg|cg|gc||at|cg|at||cg|ta|gc||gc|gc|gc||a&quot;| __truncated__</code></pre>
<ul>
<li><code>couplet_amino</code>: a pattern in which every codon (i.e. 3 base pairs) corresponding to an amino acid gets its own couplet, where the initial nucleotide from one side of the base pair gets the first line of the couplet, and the corresponding nucleotides of the base pairs get the second line of the couplet. This creates a pattern where there are three patterned words in each line. <em>Similar to <code>tercet_amino</code>, but turned sideways.</em></li>
</ul>
<pre class="r"><code>str(vision_patterns$couplet_amino$pattern)</code></pre>
<pre><code>##  chr &quot;cac|gtg||tgg|acc||gga|cct||acg|tgc||tct|aga||tgg|acc||gat|cta||ccg|ggc||aca|tgt||ctg|gac||ggg|ccc||aac|ttg||gtc&quot;| __truncated__</code></pre>
</div>
<div id="step-5-fill-the-pattern-with-words" class="section level2">
<h2>Step 5: Fill the Pattern with Words</h2>
<p>Finally, we need to replace the nucleotides (A, T, C, G) with some verbal/written patterns to use in the poem. So, we define patterns of words to stand in for each nucleotide. Some options for the types of word patterns we might use, based on their shared characteristics, include:</p>
<ul>
<li><code>letter_proportion</code>: words which have a certain proportion of their characters sharing the same letter</li>
<li><code>vocal_proportion</code>: a concentration of similar vocal pattern (such as fricatives)</li>
<li><code>cluster</code>: clustered together by similarity across multiple metrics, using <em>kmeans</em></li>
<li><code>rhyme</code>: rhyme based on Soundex or other algorithms</li>
<li><code>category</code>: words from a given category (e.g. using the <em>Regressive Imagery Dictionary</em> classifications)</li>
<li><code>regex</code>: matches a regular expression (e.g. any word beginning with <code>w</code> and containing an <code>o</code> or <code>u</code> in the middle before ending with the letter <code>d</code>).</li>
<li>Various other potential methods.</li>
</ul>
</div>
<div id="appendix-other-options-for-packages" class="section level2">
<h2>Appendix: Other options for Packages</h2>
<p>If you’re a fan of this kind of thing, there are various other options to read in the human genome for use in constructing patterns, each with their own pros and cons. Here are a few:</p>
<ul>
<li>The <a href="10.18129/B9.bioc.org.Hs.eg.db">org.Hs.eg.db</a> package <em>Genome wide annotation for Human</em>, primarily based on mapping using Entrez Gene identifiers. This requires working with the data as a <code>biostrings</code> object, but does include association data which will be helpful in locating the nucleotide sequences for use in specific portions of the poem.</li>
<li>The <a href="10.18129/B9.bioc.BSgenome.Hsapiens.UCSC.hg19">BSgenome.Hsapiens.UCSC.hg19</a> <em>Full genome sequences for Homo sapiens (UCSC version hg19)</em></li>
<li>The <a href="https://ropensci.github.io/biomartr/index.html"><code>biomartr</code></a> package<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>, which allows for retrieval of all available kingdoms stored in the <code>NCBI RefSeq</code>, <code>NCBI Genbank</code>, <code>ENSEMBL</code>, and <code>ENSEMBLGENOMES</code> sources. It also allows for download as either <code>biostrings</code> or <code>data.table</code> object.</li>
<li>The <a href="https://github.com/ropensci/rentrez"><code>rentrez</code></a> package, which allows querying the NCBI Entrez Eutils API to search, download data from, and interact with databases.</li>
</ul>
</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>Carlson M (2018). GO.db: A set of annotation maps describing the entire Gene Ontology. R package version 3.6.0. DOI: <a href="10.18129/B9.bioc.GO.db">10.18129/B9.bioc.GO.db</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>For this, we use the <code>hgu133plus2.db</code> package, which allows us to query by any of the following fields: ACCNUM, ALIAS, ENSEMBL, ENSEMBLPROT, ENSEMBLTRANS, ENTREZID, ENZYME, EVIDENCE, EVIDENCEALL, GENENAME, GENETYPE, GO, GOALL, IPI, MAP, OMIM, ONTOLOGY, ONTOLOGYALL, PATH, PFAM, PMID, PROBEID, PROSITE, REFSEQ, SYMBOL, UCSCKG, UNIPROT. We’ll be using the <code>GO</code> field to match our <code>GOID</code>s to additional information. Notes from this site on <a href="http://genomicsclass.github.io/book/pages/bioc1_annoOverview.html">using GO a.k.a. gene ontology</a> for identifying functional areas is helpful, as is <a href="https://www.bioconductor.org/help/course-materials/2015/UseBioconductorFeb2015/A01.5_Annotation.html">this vignette on the Bioconductor site</a>.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>In the current code, I’m pulling the <em>Affymetrix Human Genome U133 Plus 2.0 Array annotation data</em> using the <a href="http://bioconductor.org/packages/release/data/annotation/html/hgu133plus2.db.html"><code>hgu133plus2.db</code> package</a>.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>For instance, we may want to filter on the strength of evidence. Based on the GO documentation “IEA is a weak association and is based on electronic information, no human curator has examined or confirmed this association… IEA is also the most common evidence code.” Which speaks to the tip-of-the-iceberg nature or our current knowledge.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>Drost HG, Paszkowski J. Biomartr: genomic data retrieval with R. <em>Bioinformatics</em> (2017) 33(8): 1216-1217. <a href="doi:10.1093/bioinformatics/btw821" class="uri">doi:10.1093/bioinformatics/btw821</a><a href="#fnref5" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
